mautic/mautic
# Form object refactoring

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:


This PR replaces [PR #11234](https://github.com/mautic/mautic/pull/11234): rebased and errors fixes

#### Steps to test this PR:


1. Open [mautic/mautic on Gitpod](https://gitpod.io/#https://github.com/mautic/mautic) or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester)) - as you need to do some things before applying the PR.
1. Before you checkout this PR:
2. Create a form with at least 4 fields. One mapped to the contact email field, one mapped to the company field under the contact opt group, one mapped to the primary company city field. The fourth field should not be mapped to any field.
3. Check out this PR - in Gitpod on the left hand side, look for the Github icon, then find this PR and right click to checkout the PR:
![screenshot-mautic-mautic-oq31hzs9992 ws-eu85 gitpod io-2023 02 04-15_36_18](https://user-images.githubusercontent.com/2930593/216773288-7d0610a4-f7b8-45c9-a2fb-4dffdc7232d2.png)

5. Run the migration using the terminal tab. 

Run `ddev exec bin/console doctrine:migrations:execute --up 20200415135706`. Note:  you can roll back your changes with `ddev exec bin/console doctrine:migrations:execute --down 20200415135706` if needed.

6. Open PHPMyAdmin under the Ports tab on Gitpod, look for the port 8036. Open it in a tab.

7. Open the database named 'db' and go to the `form_fields` table and check that the last 2 columns are `mapped_object` and `mapped_field`. The field that was mapped to the contact email field should have `mapped_object = 'lead'` and `mapped_field = 'email'`. The one that was mapped to the company field under the contact opt group should have `mapped_object = 'lead'` and `mapped_field = 'company'`. The one that was mapped to the primary company city field should have `mapped_object = 'company'` and `mapped_field = 'companycity'`. The form field that was not mapped at all should have null values there.
8. When you map one field then it should be removed from mapping of other fields. Example: Map contact email field to a form field. Create another field. The contact email field is not in the mapping list.
9. Test submissions of different field types.
10. Test conditional fields
11. Test progressive profiling (behaviours)

#### List deprecations along with the new alternative:
1. `form_fields.lead_field` column is deprecated. Use `mapped_object` and `mapped_field` instead.
2. `Mautic\FormBundle\Entity\Field::getLeadField()` and `setLeadField()`. Use getters and setters on `mappedObject` and `mappedField` properties instead.
3. `Mautic\FormBundle\Event\Service\FieldValueTransformer::isIsTransformed()` is deprecated as not used.



<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11905"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
